src/nvim/terminal.c

```diff
@@ -313,6 +313,10 @@ void terminal_resize(Terminal *term, uint16_t width, uint16_t height)
     return;
   }
 
+  if (height == 0 || width == 0) {
+    return;
+  }
```

Comment:
> Just for the record, normally, I'd request that such changes be put in the main PR, but I do prefer to get bug fixes merged sooner than later, and it's only one commit.

In the other PR there's some more debate going on, so I didn't want to add yet another commit to it, since it's a bugfix. Throwing those together would probably delay both the bugfix and the new behaviour I'd like to introduce. I was hoping to get this merged in rather quickly.

As for analysis of the problem : 

The reason we end up with a 0 height is the following : when splitting a window (`win_split_ins` in `window.c`), we first set the new width (`win_new_width`) which updates the terminal via `terminal_resize` (but at that time the height hasn't been set yet). After that we set the new height (`win_new_height`) for the one window, then for the other. This function also calls `terminal_resize`, with more correct values.

With a breakpoint at `terminal_resize`, doing a `:sp` and each time printing the stack and the arguments I get the following : 

```
Breakpoint 1, terminal_resize (term=0x7fcd28b0a000, width=80, height=height@entry=0)
    at /home/frederik/neovim/neovim/src/nvim/terminal.c:286
286	{
(gdb) where
#0  terminal_resize (term=0x7fcd28b0a000, width=80, height=height@entry=0)
    at /home/frederik/neovim/neovim/src/nvim/terminal.c:286
#1  0x000000000047ebfd in win_new_width (wp=wp@entry=0x7fcd28925000, 
    width=<optimized out>) at /home/frederik/neovim/neovim/src/nvim/window.c:4748
#2  0x0000000000482ce8 in win_split_ins (size=size@entry=0, flags=flags@entry=0, 
    new_wp=new_wp@entry=0x0, dir=dir@entry=0)
    at /home/frederik/neovim/neovim/src/nvim/window.c:894
#3  0x00000000004831ad in win_split (size=0, flags=0)
    at /home/frederik/neovim/neovim/src/nvim/window.c:535
#4  0x00000000004919cb in ex_splitview (eap=0x7fffbdaac6f0)
    at /home/frederik/neovim/neovim/src/nvim/ex_docmd.c:6265
#5  0x000000000048fb30 in do_one_cmd (cmdlinep=cmdlinep@entry=0x7fffbdaac840, 
    sourcing=sourcing@entry=0, cstack=cstack@entry=0x7fffbdaac8d0, 
    fgetline=fgetline@entry=0x4ba7e5 <getexline>, cookie=cookie@entry=0x0)
    at /home/frederik/neovim/neovim/src/nvim/ex_docmd.c:2196
#6  0x0000000000490492 in do_cmdline (cmdline=<optimized out>, cmdline@entry=0x0, 
    fgetline=0x4ba7e5 <getexline>, cookie=cookie@entry=0x0, flags=0)
    at /home/frederik/neovim/neovim/src/nvim/ex_docmd.c:622
#7  0x000000000049bcc8 in nv_colon (cap=0x7fffbdaace00)
    at /home/frederik/neovim/neovim/src/nvim/normal.c:4078
#8  0x00000000004a42ac in normal_cmd (oap=oap@entry=0x7fffbdaacea0, 
    toplevel=toplevel@entry=true) at /home/frederik/neovim/neovim/src/nvim/normal.c:911
#9  0x00000000004f600e in main_loop (cmdwin=cmdwin@entry=0, noexmode=noexmode@entry=0)
    at /home/frederik/neovim/neovim/src/nvim/main.c:706
#10 0x00000000004f6f32 in main (argc=<optimized out>, argv=<optimized out>)
    at /home/frederik/neovim/neovim/src/nvim/main.c:493
(gdb) info args
term = 0x7fcd28b0a000
width = 80
height = 0
(gdb) c
Continuing.

Breakpoint 1, terminal_resize (term=0x7fcd28b0a000, width=width@entry=0, height=11)
    at /home/frederik/neovim/neovim/src/nvim/terminal.c:286
286	{
(gdb) where
#0  terminal_resize (term=0x7fcd28b0a000, width=width@entry=0, height=11)
    at /home/frederik/neovim/neovim/src/nvim/terminal.c:286
#1  0x000000000047f11f in win_new_height (wp=wp@entry=0x7fcd28925000, 
    height=height@entry=11) at /home/frederik/neovim/neovim/src/nvim/window.c:4726
#2  0x0000000000482d09 in win_split_ins (size=size@entry=0, flags=flags@entry=0, 
    new_wp=new_wp@entry=0x0, dir=dir@entry=0)
    at /home/frederik/neovim/neovim/src/nvim/window.c:901
#3  0x00000000004831ad in win_split (size=0, flags=0)
    at /home/frederik/neovim/neovim/src/nvim/window.c:535
#4  0x00000000004919cb in ex_splitview (eap=0x7fffbdaac6f0)
    at /home/frederik/neovim/neovim/src/nvim/ex_docmd.c:6265
#5  0x000000000048fb30 in do_one_cmd (cmdlinep=cmdlinep@entry=0x7fffbdaac840, 
    sourcing=sourcing@entry=0, cstack=cstack@entry=0x7fffbdaac8d0, 
    fgetline=fgetline@entry=0x4ba7e5 <getexline>, cookie=cookie@entry=0x0)
    at /home/frederik/neovim/neovim/src/nvim/ex_docmd.c:2196
#6  0x0000000000490492 in do_cmdline (cmdline=<optimized out>, cmdline@entry=0x0, 
    fgetline=0x4ba7e5 <getexline>, cookie=cookie@entry=0x0, flags=0)
    at /home/frederik/neovim/neovim/src/nvim/ex_docmd.c:622
#7  0x000000000049bcc8 in nv_colon (cap=0x7fffbdaace00)
    at /home/frederik/neovim/neovim/src/nvim/normal.c:4078
#8  0x00000000004a42ac in normal_cmd (oap=oap@entry=0x7fffbdaacea0, 
    toplevel=toplevel@entry=true) at /home/frederik/neovim/neovim/src/nvim/normal.c:911
#9  0x00000000004f600e in main_loop (cmdwin=cmdwin@entry=0, noexmode=noexmode@entry=0)
    at /home/frederik/neovim/neovim/src/nvim/main.c:706
#10 0x00000000004f6f32 in main (argc=<optimized out>, argv=<optimized out>)
    at /home/frederik/neovim/neovim/src/nvim/main.c:493
(gdb) info args
term = 0x7fcd28b0a000
width = 0
height = 11
(gdb) c    
Continuing.

Breakpoint 1, terminal_resize (term=0x7fcd28b0a000, width=width@entry=0, height=10)
    at /home/frederik/neovim/neovim/src/nvim/terminal.c:286
286	{
(gdb) where
#0  terminal_resize (term=0x7fcd28b0a000, width=width@entry=0, height=10)
    at /home/frederik/neovim/neovim/src/nvim/terminal.c:286
#1  0x000000000047f11f in win_new_height (wp=wp@entry=0x7fcd2887e000, 
    height=height@entry=10) at /home/frederik/neovim/neovim/src/nvim/window.c:4726
#2  0x0000000000482d3c in win_split_ins (size=size@entry=0, flags=flags@entry=0, 
    new_wp=new_wp@entry=0x0, dir=dir@entry=0)
    at /home/frederik/neovim/neovim/src/nvim/window.c:906
#3  0x00000000004831ad in win_split (size=0, flags=0)
    at /home/frederik/neovim/neovim/src/nvim/window.c:535
#4  0x00000000004919cb in ex_splitview (eap=0x7fffbdaac6f0)
    at /home/frederik/neovim/neovim/src/nvim/ex_docmd.c:6265
#5  0x000000000048fb30 in do_one_cmd (cmdlinep=cmdlinep@entry=0x7fffbdaac840, 
    sourcing=sourcing@entry=0, cstack=cstack@entry=0x7fffbdaac8d0, 
    fgetline=fgetline@entry=0x4ba7e5 <getexline>, cookie=cookie@entry=0x0)
    at /home/frederik/neovim/neovim/src/nvim/ex_docmd.c:2196
#6  0x0000000000490492 in do_cmdline (cmdline=<optimized out>, cmdline@entry=0x0, 
    fgetline=0x4ba7e5 <getexline>, cookie=cookie@entry=0x0, flags=0)
    at /home/frederik/neovim/neovim/src/nvim/ex_docmd.c:622
#7  0x000000000049bcc8 in nv_colon (cap=0x7fffbdaace00)
    at /home/frederik/neovim/neovim/src/nvim/normal.c:4078
#8  0x00000000004a42ac in normal_cmd (oap=oap@entry=0x7fffbdaacea0, 
    toplevel=toplevel@entry=true) at /home/frederik/neovim/neovim/src/nvim/normal.c:911
#9  0x00000000004f600e in main_loop (cmdwin=cmdwin@entry=0, noexmode=noexmode@entry=0)
    at /home/frederik/neovim/neovim/src/nvim/main.c:706
#10 0x00000000004f6f32 in main (argc=<optimized out>, argv=<optimized out>)
    at /home/frederik/neovim/neovim/src/nvim/main.c:493
(gdb) info args
term = 0x7fcd28b0a000
width = 0
height = 10
```
If you like, I can add this explanation to the commit message.

Concerning doing it in the loop, is it something like this you have in mind? 
```
void terminal_resize(Terminal *term, uint16_t width, uint16_t height)
{
  if (term->closed) {
    // will be called after exited if two windows display the same terminal and
    // one of the is closed as a consequence of pressing a key.
    return;
  }
  int curwidth, curheight;
  uint16_t width_temp, height_temp;
  vterm_get_size(term->vt, &curheight, &curwidth);

  if (!width) {
    width = (uint16_t)curwidth;
  }

  if (!height) {
    height = (uint16_t)curheight;
  }

  // The new width/height are the minimum for all windows that display the
  // terminal in the current tab.
  FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
    if (!wp->w_closing && wp->w_buffer == term->buf) {
      width_temp = (uint16_t)MIN(width,
          (uint16_t)(wp->w_width - win_col_off(wp)));
      if (width_temp != 0) {
        width = width_temp;
      }
      height_temp = (uint16_t)MIN(height, (uint16_t)wp->w_height);
      if (height_temp != 0) {
        height = height_temp;
      }
    }
  }

  if (curheight == height && curwidth == width) {
    return;
  }

  vterm_set_size(term->vt, height, width);
  vterm_screen_flush_damage(term->vts);
  term->pending_resize = true;
  invalidate_terminal(term, -1, -1);
}
```
I'm not really sure it makes all that much difference, either fix is fine by me :-)